servicetalk-http-utils/src/main/java/io/servicetalk/http/utils/RedirectSingle.java

```diff
@@ -184,14 +185,19 @@ public void onSuccess(@Nullable final StreamingHttpResponse response) {
                     LOGGER.trace("Executing redirect to '{}' for request '{}'", location, request);
                 }
 
+                // Consume any payload of the redirect response
+                final Single<StreamingHttpResponse> nextResponse = response.messageBody().ignoreElements()
+                        .concat(redirectSingle.requester.request(newRequest));
+                final RedirectSubscriber redirectSubscriber = new RedirectSubscriber(target, redirectSingle, newRequest,
+                        redirectCount + 1, sequentialCancellable);
                 terminalDelivered = true;   // Mark as "delivered" because we do not own `target` from this point
-                toSource(response.messageBody().ignoreElements()    // Consume any payload of the redirect response
-                        .concat(redirectSingle.requester.request(newRequest)))
-                        .subscribe(new RedirectSubscriber(target, redirectSingle, newRequest, redirectCount + 1,
-                                sequentialCancellable));
+                toSource(nextResponse).subscribe(redirectSubscriber);
             } catch (Throwable cause) {
                 if (!terminalDelivered) {
-                    safeOnError(target, cause);
+                    // Drain response payload body before propagating the cause
+                    sequentialCancellable.nextCancellable(response.messageBody().ignoreElements()
+                            .whenOnError(suppressed -> safeOnError(target, addSuppressed(cause, suppressed)))
+                            .subscribe(() -> safeOnError(target, cause)));
                 } else {
                     LOGGER.info("Ignoring exception from onSuccess of Subscriber {}.", target, cause);
```

Comment:
In the `terminalDelivered == true` path we still don't drain and I see that we set that flag to true _before_ we call many of the callbacks, eg
```
terminalDelivered = true;
target.onSuccess(response);
```
If it is `target.onSuccess(..)` that threw we will potentially still leak unless the `onSuccess` call did the draining. That does seem like it should be the responsibility of the `onSucess` method, but how defensive do we want to be?

Reply:
The logic here is that we set the `terminalDelivered = true` and in the next line we subscribe to `nextResponse` single, which will first subscribe to `response.messageBody().ignoreElements()`, register its cancellable in `sequentialCancellable`, and only then will deal with `target`. This guarantees that we do not leak the response, it will either terminate or get canceled.

Reply:
I think we may be talking about different cases as I'm talking about those higher up which you didn't modify. For example:
```
                final String location = redirectLocation(redirectCount, request, response);
                if (location == null) {
                    terminalDelivered = true;
                    target.onSuccess(response); // <- if I throw, is the response body drained or not?
                    return;
                }
```
There are a few others that follow the same pattern. Trying to determine if the body has been drained is perhaps too defensive but we also don't normally expect `onSuccess(..)` to throw so maybe it's better to aggressively drain?

Reply:
Change otherwise LGTM, but I'd also like to understand how the flow works in the case that @bryce-anderson outlined.

Reply:
Good point, if users have an unexpected exception inside an operator applied later (for example, `Single.map(...)`), there is a risk they can leak the response as well. 
First, users code is outside of our responsibility. They should not throw from operators but propagate cancel/onError via reactive flow. 
Second, as we discussed offline we can provide a general filter for all clients to handle this type of problem with the best effort. I will open a follow-up for that.